libtfhe/core/arithmetic/polynomial_big.cpp

```diff
@@ -0,0 +1,256 @@
+}
+
+template<typename TYPE>
+void Polynomial<TYPE>::Copy(
+    Polynomial<TYPE> *result,
+    const Polynomial<TYPE> *source,
+    const PolynomialParams<TYPE> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    assert(result != source);
+
+    const int32_t N = params->N;
+    const TYPE *s = source->coefs;
+    TYPE *r = result->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        copy(r+i, s+i, params->zmodule_params);
+}
+
+template<typename TYPE>
+void Polynomial<TYPE>::Add(
+    Polynomial<TYPE> *result,
+    const Polynomial<TYPE> *poly1,
+    const Polynomial<TYPE> *poly2,
+    const PolynomialParams<TYPE> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    assert(result != poly1); //if it fails here, please use addTo
+    assert(result != poly2); //if it fails here, please use addTo
+
+    const int32_t N = params->N;
+    TYPE *r = result->coefs;
+    const TYPE *a = poly1->coefs;
+    const TYPE *b = poly1->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        add(r+i, a+i, b+i, params->zmodule_params);
+}
+
+template<typename TYPE>
+void Polynomial<TYPE>::AddTo(
+    Polynomial<TYPE> *result,
+    const Polynomial<TYPE> *poly,
+    const PolynomialParams<TYPE> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    assert(result != poly);
+
+    const int32_t N = params->N;
+    const TYPE *s = poly->coefs;
+    TYPE *r = result->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        add(r+i, r+i, s+i, params->zmodule_params);
+}
+
+template<typename TYPE>
+void Polynomial<TYPE>::Sub(
+    Polynomial<TYPE> *result,
+    const Polynomial<TYPE> *poly1,
+    const Polynomial<TYPE> *poly2,
+    const PolynomialParams<TYPE> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    assert(result != poly1); //if it fails here, please use addTo
+    assert(result != poly2); //if it fails here, please use addTo
+
+    const int32_t N = params->N;
+    TYPE *r = result->coefs;
+    const TYPE *a = poly1->coefs;
+    const TYPE *b = poly1->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        sub(r+i, a+i, b+i, params->zmodule_params);
+}
+
+template<typename TYPE>
+void Polynomial<TYPE>::SubTo(
+    Polynomial<TYPE> *result,
+    const Polynomial<TYPE> *poly,
+    const PolynomialParams<TYPE> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    assert(result != poly);
+
+    const int32_t N = params->N;
+    const TYPE *s = poly->coefs;
+    TYPE *r = result->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        sub(r+i, r+i, s+i, params->zmodule_params);
+}
+
+template<typename TYPE>
+void Polynomial<TYPE>::MulByXaiMinusOne(
+    Polynomial<TYPE> *result,
+    const int32_t a,
+    const Polynomial<TYPE> *source,
+    const PolynomialParams<TYPE> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    const int32_t N = params->N;
+    TYPE *out = result->coefs;
+    const TYPE *in = source->coefs;
+    const auto *zparams =
```

Comment:
I prefer not to use auto if the scope of the variable is > 5 lines.

Reply:
added a typedef in PolynomialParams